connectors/sources/salesforce.py

```diff
@@ -177,6 +188,96 @@ async def close(self):
         await self.session.close()
         del self.session
 
+    def modify_soql_query(self, query):
+        lowered_query = query.lower()
+        match_limit = re.search(r"(?i)(.*)FROM\s+(.*?)(?:LIMIT)(.*)", lowered_query)
+        match_offset = re.search(r"(?i)(.*)FROM\s+(.*?)(?:OFFSET)(.*)", lowered_query)
+
+        if "fields" in lowered_query and not match_limit:
+            query += " LIMIT 200 OFFSET 0"
+
+        elif "fields" in lowered_query and match_limit and not match_offset:
+            query += " OFFSET 0"
+
+        elif "fields" in lowered_query and match_limit and match_offset:
+            return query
+
+        return query
+
+    async def get_sync_rules_results(self, rule):
+        def _add_last_modified_date(query):
+            lowered_query = query.lower()
+            if (
+                not (
+                    "fields(all)" in lowered_query
+                    or "fields(standard)" in lowered_query
+                )
+                and "lastmodifieddate" not in lowered_query
+            ):
+                query = re.sub(
+                    r"(?i)SELECT (.*) FROM", r"SELECT \1, LastModifiedDate FROM", query
+                )
+
+            return query
+
+        def _add_id(query):
+            lowered_query = query.lower()
+            if not (
+                "fields(all)" in lowered_query or "fields(standard)" in lowered_query
+            ) and not re.search("\bid\b", lowered_query):
+                query = re.sub(r"(?i)SELECT (.*) FROM", r"SELECT \1, Id FROM", query)
+
+            return query
+
+        if rule["language"] == "SOQL":
+            query_with_id = _add_id(query=rule["query"])
+            query = _add_last_modified_date(query=query_with_id)
+
+            if "fields" not in query.lower():
+                async for records in self._yield_non_bulk_query_pages(soql_query=query):
+                    for record in records:
+                        yield record
+            # If FIELDS function is present in SOQL query, LIMIT/OFFSET is used for pagination
+            else:
+                soql_query = self.modify_soql_query(query=query)
+                async for records in self._yield_soql_query_pages_with_fields_function(
+                    soql_query=soql_query
+                ):
+                    for record in records:
+                        yield record
+
+        else:
```

Comment:
Should this be an `elif` to check for `sosl`, and then add an `else` that raises if it's neither `soql` or `sosl`?